### DISCUSSION TITLE
Moonray 1.7.0 MacOS Metal Support

### DISCUSSION THREAD

**Discussion by jujoje** [dreamworksanimation/openmoonray#175] [Q&A]
Hi,

Built the new version of Moonray on Sonoma, and all went well (the new instructions for getting it going with Houdini were pretty good!). 

One thing I'm finding is that the xPU mode appears to be only using the CPU, not the GPU: I've tried rendering the ALAB scene in Houdini, as well as the curve test scene from the command line and neither of them seem to be going to the GPU at all.

Is there anything additional I have to do? The previous version seemed to work with Metal out of the box (the curve scene was using about 75% of the GPU iirc).

Cheers,

Julian

**Comment by randypacker** [comment]
Hi Julian,

First, so we can check, are you running explicit xpu mode, like this?

```console
moonray_gui -exec_mode xpu -info -in curves.rdla
```

We've built 1.7 on Sonoma and can see the GPU access also at about 75% on that curves test scene when run in xpu mode.  It should also be working for you, if you could confirm.  If not, we should investigate further.

Hope this helps!  Let us know, thanks.

**Reply by jujoje** [reply]
Hi Randy,

Thanks for the reply!

Tried the curve scene with those options and it looks like it _is_ actually using the GPU; from the logs it is detecting it find and using it for the render. However it looks like it's only using 15-20% of GPU (at least according to Activity Monitor).

SLso I just double checked the machine I was running this on and looks it had been updated to Sequoia (15.1.1). Wonder if it might be down the OS update; 1.6 worked fine on Sonoma. (other than that it was a Mac Studio Ultra M2)

**Reply by jmahovsky-dwa** [reply]
Hi Julian,

If you are running XPU on Mac/Metal, it is limited to accelerating the occlusion (light sampling) rays, and is hence underutilized.  The regular Linux/Optix XPU accelerates both regular rays and occlusion rays, enabling more speedup.  We will be supporting regular rays on Mac XPU in a future release as we still need to port all that new code to the Metal API.

If you benchmark the curves scene in XPU mode versus running in vector mode, you may find that XPU mode is actually slower than vector mode.  This is a known issue when running XPU mode with scenes with very simple geometry.  Basically, the GPU doesn't have much work to do per-ray to find the intersections, so not much acceleration is possible, but there is still the same overhead of copying rays to the GPU and copying intersection results back from the GPU.  So, the XPU acceleration is insufficient to overcome the additional XPU overhead.

We find that XPU performs best when rendering scenes with dense, complex geometry (like hair) or foliage.  It takes a lot more work to find the ray intersections in these scenes so the GPU has more to do.

The long-term plan is to offload more/different kinds of work onto the GPU to give it a wider variety of work and faster/more balanced performance increase.
